Homework for sprint №13

sobolevkir opened the pull request.

sobolevkir committed:
Player activity have been added.
It appears when click on track.

vlad-zotov commented on app/src/main/java/com/sobolevkir/playlistmaker/PlayerActivity.kt:
```diff
@@ -0,0 +1,97 @@
+
+class PlayerActivity : AppCompatActivity() {
+
+    private lateinit var backButton: ImageButton
```
Привет! lateinit var не рекомендуется использовать, из-за того, что переменная будет иметь не нуллабельный тип но при этом может быть не инициализирована. Про это можно забыть при написании кода и получить падение в непредвиденный момент.

sobolevkir replied:
а что тогда сделать лучше?

vlad-zotov replied:
делать private var backButton: ImageButton? и обращаться как к null-абельному типу

vlad-zotov commented on app/src/main/java/com/sobolevkir/playlistmaker/PlayerActivity.kt:
```diff
@@ -0,0 +1,97 @@
+        backButton.setOnClickListener { finish() }
+
+        val track = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            intent.getParcelableExtra("current_track", Track::class.java)
```
советую выносить строки-ключи в константы. Так ты сможешь переиспользовать их в проекте, а не копировать все время одинаковые строки типа `"current_track"`. Копирование кода ведет к проблемам при его поддержке и переписывании.

sobolevkir replied:
где лучше эту константу вставить? подскажи пожалуйста конкретное место! Спасибо!

vlad-zotov replied:
в `companion object` этого класса

vlad-zotov commented on app/src/main/java/com/sobolevkir/playlistmaker/PlayerActivity.kt:
```diff
@@ -0,0 +1,97 @@
+        if ((track?.collectionName != null) && track.collectionName.isNotEmpty()) {
+            collectionName.text = track.collectionName
+        }
+        else {
```
по стилю Kotlin else должен быть на одной строке с предыдущей закрывающей скобкой `{`

vlad-zotov commented on app/src/main/java/com/sobolevkir/playlistmaker/PlayerActivity.kt:
```diff
@@ -0,0 +1,97 @@
+        if ((track?.releaseDate != null) && track.releaseDate.isNotEmpty()) {
+            releaseYear.text = track.getReleaseYear()
+        }
+        else {
```
в Java/Kotlin принят немного другой стиль форматирования кода, легко и быстро отформатировать код в Android Studio можно следующей комбинацией клавиш: в Windows Ctrl + Alt + L , в MacOs ⌘ + ⌥ + L

vlad-zotov commented on app/src/main/java/com/sobolevkir/playlistmaker/PlayerActivity.kt:
```diff
@@ -0,0 +1,97 @@
+
+        genre = findViewById(R.id.tv_genre_value)
+        genreTitle = findViewById(R.id.tv_genre_title)
+        if ((track?.primaryGenreName != null) && track.primaryGenreName.isNotEmpty()) {
```
можно написать проще, используя kotlin extension функцию `if(!track?.primaryGenreName.isNullOrEmpty())`

sobolevkir replied:
Да, точно, вспомнил про неё.
Подскажи пожалуйста, а почему в данном случае нужно безопасное присваивание делать через "?"
Ведь мы уже сделали проверку в условии на null-значение...

vlad-zotov replied:
`track?.` проверка на нулл для трека. Получается, что если значением `track` является `null`, то `primaryGenreName` мы не сможем получить то есть оно тоже будет `null` и работать с этим полем нужно как с null-абельным.

vlad-zotov commented on app/src/main/java/com/sobolevkir/playlistmaker/PlayerActivity.kt:
```diff
@@ -0,0 +1,97 @@
+
+        country = findViewById(R.id.tv_country_value)
+        countryTitle = findViewById(R.id.tv_country_title)
+        if ((track?.country != null) && track.country.isNotEmpty()) {
```
здесь тоже можно упростить, используя `isNullOrEmpty()`

vlad-zotov commented on app/src/main/java/com/sobolevkir/playlistmaker/tracklist/Track.kt:
```diff
@@ -1,8 +1,24 @@
+    val collectionName: String?,
+    val releaseDate: String?,
+    val primaryGenreName: String?,
+    val country: String?
```
советую ставить в конце последнего параметра запутяю, так называемую `trailing comma`. Такой подход позволяет уменьшить число измененных строк в диффах git, когда проверяешь реквесты. Потому что будет добавлена только одна строка и не будет изменена верхнаяя. Обрати внимание как выглядит твой дивв на 7-8 строках этого файла.

vlad-zotov commented on app/src/main/java/com/sobolevkir/playlistmaker/tracklist/TrackListAdapter.kt:
```diff
@@ -1,23 +1,31 @@
-        holder.itemView.setOnClickListener {onTrackClickListener?.onTrackClick(tracks[position])}
+        holder.itemView.setOnClickListener {
+            onTrackClickListener?.onTrackClick(tracks[position])
+            val playerIntent = Intent(it.context, PlayerActivity::class.java)
```
код запуска активити следует вынести из адаптера и выполнять в `onTrackClickListener`, который ты сюда передаёшь. Адаптер не должен быть привязан к вызову конкретной активити, это опции, которые могут меняться в процессе разработки.

vlad-zotov commented on app/src/main/java/com/sobolevkir/playlistmaker/tracklist/TrackListAdapter.kt:
```diff
@@ -1,23 +1,31 @@
+        holder.itemView.setOnClickListener {
+            onTrackClickListener?.onTrackClick(tracks[position])
+            val playerIntent = Intent(it.context, PlayerActivity::class.java)
+            playerIntent.putExtra("current_track", tracks[position])
```
строку-ключ следует вынести в константу внутри `PlayerActivity`. 
Для запуска активити в андроиде часто применяют следующий паттерн. Создают в активити `comanion object` со следующим содержанием
```
fun startActivity(context: Context) {
    val intent = Intent(context: Context, MyActivity::class.java)
    intent.putExtra(EXTRA_KEY, value)
    context.startActivity(intent)
}
```

vlad-zotov commented on app/src/main/res/drawable-night/btn_add_icon.xml:
```diff
@@ -0,0 +1,13 @@
+    android:viewportHeight="51">
+  <path
+      android:pathData="M25.5,0L25.5,0A25.5,25.5 0,0 1,51 25.5L51,25.5A25.5,25.5 0,0 1,25.5 51L25.5,51A25.5,25.5 0,0 1,0 25.5L0,25.5A25.5,25.5 0,0 1,25.5 0z"
+      android:fillColor="#ffffff"
```
следует выносить цвета в ресуры

vlad-zotov commented on app/src/main/res/drawable-night/btn_add_icon.xml:
```diff
@@ -0,0 +1,13 @@
+      android:fillAlpha="0.25"/>
+  <path
+      android:pathData="M16.3,18.6H14V34.7C14,35.965 15.035,37 16.3,37H32.4V34.7H16.3V18.6ZM34.7,14H20.9C19.635,14 18.6,15.035 18.6,16.3V30.1C18.6,31.365 19.635,32.4 20.9,32.4H34.7C35.965,32.4 37,31.365 37,30.1V16.3C37,15.035 35.965,14 34.7,14ZM33.55,24.35H28.95V28.95H26.65V24.35H22.05V22.05H26.65V17.45H28.95V22.05H33.55V24.35Z"
+      android:fillColor="#ffffff"/>
```
следует выносить цвета в ресуры

vlad-zotov commented on app/src/main/res/drawable-night/btn_like_icon.xml:
```diff
@@ -0,0 +1,13 @@
+    android:viewportHeight="51">
+  <path
+      android:pathData="M25.5,0L25.5,0A25.5,25.5 0,0 1,51 25.5L51,25.5A25.5,25.5 0,0 1,25.5 51L25.5,51A25.5,25.5 0,0 1,0 25.5L0,25.5A25.5,25.5 0,0 1,25.5 0z"
+      android:fillColor="#ffffff"
```
следует выносить цвета в ресуры

vlad-zotov commented on app/src/main/res/drawable-night/btn_like_icon.xml:
```diff
@@ -0,0 +1,13 @@
+      android:fillAlpha="0.25"/>
+  <path
+      android:pathData="M31.125,15C28.95,15 26.862,16.015 25.5,17.62C24.138,16.015 22.05,15 19.875,15C16.025,15 13,18.033 13,21.894C13,26.632 17.25,30.492 23.688,36.358L25.5,38L27.313,36.346C33.75,30.492 38,26.632 38,21.894C38,18.033 34.975,15 31.125,15ZM25.625,34.491L25.5,34.616L25.375,34.491C19.425,29.088 15.5,25.516 15.5,21.894C15.5,19.387 17.375,17.507 19.875,17.507C21.8,17.507 23.675,18.748  ...
+      android:fillColor="#ffffff"/>
```
следует выносить цвета в ресуры

vlad-zotov commented on app/src/main/res/drawable-night/btn_play_icon.xml:
```diff
@@ -0,0 +1,9 @@
+    android:viewportHeight="100">
+  <path
+      android:pathData="M50,8.33C27,8.33 8.33,27 8.33,50C8.33,73 27,91.67 50,91.67C73,91.67 91.67,73 91.67,50C91.67,27 73,8.33 50,8.33ZM41.67,68.75V31.25L66.67,50L41.67,68.75Z"
+      android:fillColor="#ffffff"/>
```
следует выносить цвета в ресуры

vlad-zotov commented on app/src/main/res/drawable/btn_add_icon.xml:
```diff
@@ -0,0 +1,13 @@
+    android:viewportHeight="51">
+  <path
+      android:pathData="M25.5,0L25.5,0A25.5,25.5 0,0 1,51 25.5L51,25.5A25.5,25.5 0,0 1,25.5 51L25.5,51A25.5,25.5 0,0 1,0 25.5L0,25.5A25.5,25.5 0,0 1,25.5 0z"
+      android:fillColor="#1A1B22"
```
следует выносить цвета в ресуры

vlad-zotov commented on app/src/main/res/drawable/btn_add_icon.xml:
```diff
@@ -0,0 +1,13 @@
+      android:fillAlpha="0.25"/>
+  <path
+      android:pathData="M16.3,18.6H14V34.7C14,35.965 15.035,37 16.3,37H32.4V34.7H16.3V18.6ZM34.7,14H20.9C19.635,14 18.6,15.035 18.6,16.3V30.1C18.6,31.365 19.635,32.4 20.9,32.4H34.7C35.965,32.4 37,31.365 37,30.1V16.3C37,15.035 35.965,14 34.7,14ZM33.55,24.35H28.95V28.95H26.65V24.35H22.05V22.05H26.65V17.45H28.95V22.05H33.55V24.35Z"
+      android:fillColor="#ffffff"/>
```
следует выносить цвета в ресуры

vlad-zotov commented on app/src/main/res/drawable/btn_like_icon.xml:
```diff
@@ -0,0 +1,13 @@
+    android:viewportHeight="51">
+  <path
+      android:pathData="M25.5,0L25.5,0A25.5,25.5 0,0 1,51 25.5L51,25.5A25.5,25.5 0,0 1,25.5 51L25.5,51A25.5,25.5 0,0 1,0 25.5L0,25.5A25.5,25.5 0,0 1,25.5 0z"
+      android:fillColor="#1A1B22"
```
следует выносить цвета в ресуры

vlad-zotov commented on app/src/main/res/drawable/btn_like_icon.xml:
```diff
@@ -0,0 +1,13 @@
+      android:fillAlpha="0.25"/>
+  <path
+      android:pathData="M31.125,15C28.95,15 26.862,16.015 25.5,17.62C24.138,16.015 22.05,15 19.875,15C16.025,15 13,18.033 13,21.894C13,26.632 17.25,30.492 23.688,36.358L25.5,38L27.313,36.346C33.75,30.492 38,26.632 38,21.894C38,18.033 34.975,15 31.125,15ZM25.625,34.491L25.5,34.616L25.375,34.491C19.425,29.088 15.5,25.516 15.5,21.894C15.5,19.387 17.375,17.507 19.875,17.507C21.8,17.507 23.675,18.748  ...
+      android:fillColor="#ffffff"/>
```
следует выносить цвета в ресуры

vlad-zotov commented on app/src/main/res/layout/layout_tracklist_item.xml:
```diff
@@ -12,7 +12,8 @@
         android:layout_width="@dimen/album_cover_mini_size"
         android:layout_height="@dimen/album_cover_mini_size"
-        android:layout_marginStart="13dp" />
+        android:layout_marginStart="13dp"
```
размеры следует вынести в ресурсы

sobolevkir replied:
А в файле стилей тоже все размеры нужно выносить в dimens?

vlad-zotov replied:
да, всё верно. Файл `dimens` должен хранить все размеры, которые будут использоваться в верстке и стилях

vlad-zotov commented on app/src/main/res/layout/layout_tracklist_item.xml:
```diff
@@ -69,7 +70,7 @@
     <ImageView
         android:layout_width="24dp"
```
размеры следует вынести в ресурсы

sobolevkir committed:
Использован ViewBinding для PlayerActivity.
Скорректированы размеры и отступы в случае длинных названий.
Цвета и размеры вынесены в ресурсы.
